File: 103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // BFS & reverse: O(N)time & space, 31.08%, ok
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ret;
        if(!root) return ret;
        queue<TreeNode*> q;
        q.push(root);
        int level = 0;
        while(q.size()){
            level++;
            int levelSize = q.size();
            vector<int> temp;
            while(levelSize--){
                TreeNode* node = q.front();
                q.pop();
                temp.push_back(node->val);
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            if(!(level & 1)) reverse(temp.begin(), temp.end());
            ret.push_back(temp);
        }
        return ret;
    }

    // BFS without reverse: O(N)time & space, 31.08%, ok
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ret;
        if(!root) return ret;
        queue<TreeNode*> q;
        q.push(root);
        int index, level = 0;
        while(q.size()){
            level++;
            int levelSize = q.size();
            vector<int> temp(levelSize, 0);
            for(int i = 0; i < levelSize; i++){
                TreeNode* node = q.front();
                q.pop();
                index = (level & 1)? i : levelSize - i - 1;
                temp[index] = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            ret.push_back(temp);
        }
        return ret;
    }
};

File: 103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque
class Solution(object):
    
    # BFS: O(N) time & space
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root: return []
        queue, ret, left = deque([root]), [], True
        while len(queue):
            level, levelSize = [], len(queue)
            while levelSize:
                node = queue.popleft()
                level.append(node.val)
                queue += filter(None, [node.left, node.right])
                levelSize -= 1
            ret.append(level if left else level[::-1])
            left = not left
        return ret
    
    # !! two stack: O(N) time & space
    def zigzagLevelOrder(self, root):
        if not root: return []
        stack1, stack2 = [root], []
        ret, left = [], True
        while stack1 or stack2:
            level = []
            while stack1:
                node = stack1.pop()
                level.append(node.val)
                stack2 += filter(None, [node.left, node.right])
            ret.append(level)
            level = []
            while stack2:
                node = stack2.pop()
                level.append(node.val)
                stack1 += filter(None, [node.right, node.left])
            if level: ret.append(level)
        return ret
            
        